rts/rts.c

```diff
@@ -1908,7 +1919,11 @@ int main(int argc, char **argv) {
     if (ddb_host) {
         GET_RANDSEED(&seed, 0);
         log_info("Using distributed database backend replication factor of %d\n", ddb_replication);
-        db = get_remote_db(ddb_replication);
+        int random_id;
+		FASTRAND(&seed, random_id);
+		random_id = 5000 + random_id % 100;
```

Comment:
This is again just temporary for my own testing purposes with a couple of RTS nodes, it will not remain in the final PR. We will not have random ids in the final version as we cannot afford any collisions, however improbable. I will replace this with a config param.

Reply:
I like the random IDs. We can have a config option too but I think random IDs is a good default because then the user won't have to 

To avoid collisions, I think we should have a duplicate ID detection mechanism. It is the only way to be safe; leaving it to the user to specify with a configuration option does not avoid the problem, it only moves responsibility from us to the end user, which IMHO is poor UX. I want to help the user and help them prevent data loss / faulty behavior. Duplicate ID detection would be the best way to avoid that.

I don't think this needs to land in the initial PR though, we can build it on top later if you prefer.

Without knowing anything about the gossip protocol, it seems quite simple; just connect and receive information about the other nodes, check if there already is a node with our ID; if there is, exit with an error? Otherwise continue and announce ourselves?  But you'll probably tell me there is a very early hello message where we announce ourselves with our ID? ;)

Reply:
In principle it is possible to do duplicate detection of the RTS IDs using the gossip layer (or even easier, to generate unique IDs for RTSs automatically in the gossip agreement process - see below), but it would require some re-coding.
Currently IDs get propagated upon connection to the database from clients -> gossip membership agreement between servers (integrated with the existing gossip layer) -> pushed back to all other clients. In other words, a client (RTS) must have joined (and specified its ID) before it learns info about the other ones via gossip.
It is possible to have it join with essentially no ID, then have one assigned for it by the gossip / agreement layer itself automatically, so that the user has to specify nothing. I think this would be preferable to random ID picking by clients + collision detection, since the latter would be a more complicated solution achieving the same thing but requiring us to handle more corner cases (for instance there is no way to guarantee 2 RTSs joining very close in time to each other will have learned about the other one immediately via gossip, in which case we need a mechanism to change an already assigned ID. However unlikely, it is a case we still need to handle in code).
Generating the IDs during the agreement process is will be easier and require less message roundtrips, but I will do this in a separate PR, let's consider this PR v1.
For now I have added configuration params for: rts ID, rts hostname, as well as rack ID and datacenter ID. A unique ID is generated from the combination of hostname + rts ID, allowing the user to either deploy RTSs on different hosts, or several of them on the same host but with different rts IDs. In principle, if people use something like Kubernetes or other automatic framework for deployment rather than doing it by hand, it is likely they have a standard way of getting hostnames and ids dynamically upon each deployment.
But I agree it is better to have less configuration, so we can remove the rts_id parameter once we have a way to generate them automatically as above. We should still keep the rack_id and datacenter_id parameters though to support more advanced placement strategies based on various failures scenarios, such as partitions vs node failures.